tensorflow/python/estimator/canned/tensor_forest.py

```diff
@@ -0,0 +1,174 @@
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Estimator classes for TensorForest."""
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import collections
+
+from tensorflow.python.estimator import estimator
+from tensorflow.python.estimator.canned import head as head_lib
+from tensorflow.python.ops.losses import losses
+from tensorflow.python.summary import summary
+from tensorflow.python.feature_column import feature_column as feature_column_lib
+from tensorflow.python.ops import gen_tensor_forest_ops
+from tensorflow.python.ops import tensor_forest_ops, math_ops, array_ops
+
+# from tensorflow.python.util.tf_export import estimator_export
+
+_ForestHParams = collections.namedtuple('TreeHParams', [
+    'logits_dimension',
+    'n_trees', 'max_nodes', 'num_splits_to_consider',
+    'split_node_after_samples', 'is_regression',
+])
+
+VARIANCE_PREDICTION_KEY = 'prediction_variance'
+
+
+class RandomForestGraphs(object):
+  """Builds TF graphs for random forest training and inference."""
+
+  def __init__(self,
+               params,
+               configs,
+               tree_configs=None):
+    self._params = params
+    self._configs = configs
+    self._variables = tensor_forest_ops.ForestVariables(
+        self._params,
+        tree_configs=tree_configs)
+
+  def inference_graph(self, dense_features):
+    """ Builds inference graphs"""
+
+    logits = [
+        gen_tensor_forest_ops.tensor_forest_tree_predict(
+            tree_variable,
+            dense_features,
+            self._params.logits_dimension)
+        for tree_variable in self._variables
+    ]
+
+    # shape of all_predict should be [batch_size, n_trees, logits_dimension]
+    all_predict = array_ops.stack(logits, axis=1)
+    average_values = math_ops.div(
+        math_ops.reduce_sum(all_predict, 1),
+        self._params.n_trees,
+        name='probabilities')
+
+    expected_squares = math_ops.div(
+        math_ops.reduce_sum(all_predict * all_predict, 1),
+        self._params.n_trees)
+    regression_variance = math_ops.maximum(
+        0., expected_squares - average_values * average_values)
+
+    return average_values, regression_variance
+
+  def average_size(self):
+    sizes = [gen_tensor_forest_ops.tensor_forest_tree_size(tree_variable)
+             for tree_variable in self._variables]
+    return math_ops.reduce_mean(math_ops.to_float(array_ops.stack(sizes)))
+
+
+def _tf_model_fn(features,
```

Comment:
ahhh. my fault. bt works for boosted trees, but tf resembles tensorflow (also commonly abbreviated as tf). lets spell out tensor_forest_model_fn?